### PR TITLE
fix(sdk): fixes unresolved PipelineParam when static list passed to dsl.ParallelFor. Fixes #4890

### LINKED ISSUE #4890: Unresolved PipelineParam when static list passed to dsl.ParallelFor contains pipelineParam
### What steps did you take:
When a dsl.ParallelFor is created with static list containing a pipeline param, the compiler fails with error message: `RuntimeError: Internal compiler error: Found unresolved PipelineParam.`

### What happened:

You can reproduce with the following code:
```python
from kfp import dsl, components, compiler
def print_op(s: str):
    print(f"a is {s}")
    
PRINT_OP = components.func_to_container_op(print_op, base_image="python:3.7")

@dsl.pipeline(name="test pipeline")
def my_pipeline(name: str):
    
    texts = [{"val":"toot"}, {"val": str(name)}]
    
    with dsl.ParallelFor(texts) as text:
        PRINT_OP(s=text.val)

compiler_instance = compiler.Compiler()
compiler_instance.compile(pipeline_func=my_pipeline, package_path="my_pipeline.yaml")
```
Here is the error:
image

### What did you expect to happen:
I expect it to compile.

### Environment:

How did you deploy Kubeflow Pipelines (KFP)?
Kubeflow deployed on EKS

KFP version: `1.0.4`

KFP SDK version: `1.0.4`

### Anything else you would like to add:
I analysed YAML and it seems the pipeline param is not resolved inside YAML.

/kind bug
/area sdk

### PR DESCRIPTION
**Pull request by radcheb**
**Description of your changes:**

Fixes https://github.com/kubeflow/pipelines/issues/4890

More details [in this comment](https://github.com/kubeflow/pipelines/issues/4890#issuecomment-743878634)

Actually, in case of dsl.ParallelFor items is not a `PipelineParam` the compiler does not check if the python object(list) does include internally pipeline param. So they are not replaced nor taken into account in graph dependencies.

I fixed this by analysing the items in case it's not a `PipelineParam`.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?

### CODE REVIEW

**Comment by radcheb** [comment]
@chensun all tests pass in local. Do you have any clues why it breaks in CI ? The diff is a bit weired:
image

**Comment by chensun** [comment]
> @chensun all tests pass in local. Do you have any clues why it breaks in CI ? The diff is a bit weired:
> image

@radcheb Yes. I checked out your PR, and the tests pass locally for me as well.
I suspect there's something wrong with CI dependencies, will take a stab later today.

**Comment by radcheb** [comment]
Thanks @chensun for the check, I will rebase and retest.

**Comment by radcheb** [comment]
> Thanks @chensun for the check, I will rebase and retest.

All tests pass in the CI, thanks again for your support.

**Comment by scale-itx** [comment]
@chensun @Ark-kun Any updates on this PR ?

**Comment by radcheb** [comment]
@chensun @Ark-kun Any updates on this PR ?

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -521,6 +520,20 @@ def _group_to_dag_template(self, group, inputs, outputs, dependencies):
...
           # Need to sanitize the dict keys for consistency.
           loop_tasks = sub_group.loop_args.to_list_for_task_yaml()
+          nested_pipeline_params = extract_pipelineparams_from_any(loop_tasks)
+
+          # Set dependencies in case of nested pipeline_params
+          map_to_tmpl_var = {str(p): self._resolve_task_pipeline_param(p, group.type) for p in nested_pipeline_params}
+          for pipeline_param in nested_pipeline_params:
+            if pipeline_param.op_name:
+              # these pipeline_param are the output of another task
+              if 'dependencies' not in task or task['dependencies'] is None:
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py:530]: JFYI: `task.setdefault('dependencies', [])`

**Comment by Ark-kun** [comment]
Thank you for fixing this issue.
/lgtm
/approve

**Pull request merged**